tableauserverclient/server/request_factory.py

```diff
@@ -145,6 +145,34 @@ def publish_req_chunked(self, datasource_item, connection_credentials=None, conn
         return _add_multipart(parts)
 
 
+class FavoriteRequest(object):
+    def _add_to_req(self, id_, target_type, label):
+        '''
+        <favorite label="...">
+        <target_type id="..." />
+        </favorite>
+        '''
+        xml_request = ET.Element('tsRequest')
+        favorite_element = ET.SubElement(xml_request, 'favorite')
+        target = ET.SubElement(favorite_element, target_type)
+        favorite_element.attrib['label'] = label
+        target.attrib['id'] = id_
+
+        return ET.tostring(xml_request)
+
+    def add_datasource_req(self, id_, name):
+        return self._add_to_req(id_, 'datasource', name)
```

Comment:
For permissions content types, where we have a similar pattern, we use a (fake) enum for these types.
For grantees, we pull the type from the model itself with a 'tag_name' property that extracts from the XML.

I think here the former ( an enum ) is a little better than the raw strings

Reply:
Made an enum like class, and used that instead. Let me know if that is not what you intended.